diamond_blastx subworkflow

alxndrdiaz opened the pull request:
<!--
# nf-core/blobtoolkit pull request

Many thanks for contributing to nf-core/blobtoolkit!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/blobtoolkit/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/blobtoolkit/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/blobtoolkit _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

This PR is part of the Google Summer of Code 2022 project: [Conversion of the BlobToolKit pipeline to Nextflow](https://summerofcode.withgoogle.com/myprojects/details/v89lHomV), GitHub repository: https://github.com/sanger-tol/blobtoolkit

alxndrdiaz commented:
Implements diamond_blastx subworkflow, this goes next after busco subworkflow. 

1.  Moved all parameters outside from the subworflow scripts. 
2.  Fixed wrong variable definitions that are known to cause bugs. 
3. Added parameters required for testing (test.config).

muffato commented:
Can you remove the files that will be added through #28 ? e.g. busco_diamond_blastp sub-workflow, the goat module, etc. To make it clear what is specific to this sub-workflow / PR

muffato commented on conf/test.config:
```diff
@@ -41,4 +51,25 @@ process {
+         ext.args = '--mode genome'
+     }
+
+    withName: CHUNK_FASTA_BUSCO {
+          ext.args = '--chunk=100000 --overlap=0 --max-chunks=10 --min-length=1000'
+     }
+
+    withName: DIAMOND_BLASTX {
+           ext.args = '--evalue 1.0e-10 --max-target-seqs 10 --max-hsps 1'
+     }
+
+    withName: UNCHUNK_BLASTX {
+            ext.args = '--count 10'
+     }
```
Same as in #28 . This should be in `conf/modules.config`

alxndrdiaz replied:
Done.

muffato commented on modules/local/chunk_fasta_by_busco.nf:
```diff
@@ -0,0 +1,29 @@
+        --bed None 2> ${prefix}.chunk_fasta.log
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        blobtoolkit: \$( echo "3.1.0" | sed 's/blobtoolkit //g')
```
```suggestion
        blobtoolkit: \$(btk --version | cut -d' ' -f2 | sed 's/v//')
```

alxndrdiaz replied:
Done and also made the same change in `unchunk_blastx.nf` module script.

muffato commented on modules/local/chunk_fasta_by_busco.nf:
```diff
@@ -0,0 +1,29 @@
+process CHUNK_FASTA_BUSCO {
+    tag "$fasta"
+
+    container "genomehubs/blobtoolkit-blobtools"
```
Always need to specify a version
```suggestion
    container "genomehubs/blobtoolkit-blobtools:3.3.4"
```

alxndrdiaz replied:
Done.

muffato commented on modules/local/unchunk_blastx.nf:
```diff
@@ -0,0 +1,26 @@
+process UNCHUNK_BLASTX {
+    tag "$fasta"
+
+    container "genomehubs/blobtoolkit-blobtools"
```
same as above

alxndrdiaz replied:
Done

muffato commented on modules/local/unchunk_blastx.nf:
```diff
@@ -0,0 +1,26 @@
+        --out ${prefix}.reference_proteomes.out 2> ${prefix}.unchunk_blastx.log
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        blobtoolkit: \$( echo "3.1.0" | sed 's/blobtoolkit //g')
```
same as above

alxndrdiaz replied:
Done.

muffato reviewed (changes requested):
As in #28, please use `meta` throughout the pipeline

alxndrdiaz commented:
Commits associated to requested changes:

```
8140830 - alxndrdiaz, 2 hours ago : removed unrelated params
97f0275 - alxndrdiaz, 2 hours ago : input for UNCHUNK_BLASTX is now a tuple [meta, blast]
3e5a7b0 - alxndrdiaz, 2 hours ago : output is now tuple [meta, proteomes]
eaadae9 - alxndrdiaz, 2 hours ago : DIAMOND_BLASTX input is now tuple from CHUNK_FASTA_BUSCO
a4c02ce - alxndrdiaz, 2 hours ago : output is now tuple [meta, chunks]
c5b31a8 - alxndrdiaz, 2 hours ago : input is now tuple [meta, raw_proteomes]
a518519 - alxndrdiaz, 2 hours ago : CHUNK_FASTA_BUSCO input is now a tuple [meta, fasta]
7f01fc7 - alxndrdiaz, 2 hours ago : updated input to tuple [meta, fasta]
72a20d7 - alxndrdiaz, 2 hours ago : removed DIAMOND_BLASTX parameters, declared in params
bc963d8 - alxndrdiaz, 3 hours ago : added container version
0f337bd - alxndrdiaz, 3 hours ago : updated versions
c33b23e - alxndrdiaz, 3 hours ago : moved module params to modules.config
35b862e - alxndrdiaz, 3 hours ago : removed unrelated files from busco_subworkflow
```

priyanka-surana commented:
I don't understand why the base is `busco_subworkflow`, this will be added to dev eventually, once the busco subworkflow is merged.

muffato commented:
@priyanka-surana : this was my suggestion on the basis that this diamondblastx branch was started off the busco branch, so that github would only show the difference. But now that this branch has commits to remove the busco files, I think it's not necessary any more. I'll switch back to dev

muffato commented:
A bit cleaner, but it still shows all the automate-io files being removed. @alxndrdiaz : commits that remove files are not great because git really thinks you want these files removed. Could you maybe find a way of removing those commits in a rebase ?

alxndrdiaz commented on subworkflows/local/diamond_blastx.nf:
```diff
@@ -0,0 +1,55 @@
+    fasta.map { fa -> [ [id: fa.baseName ], fa ] },
+    busco_table
+    )
+    ch_versions = ch_versions.mix(CHUNK_FASTA_BUSCO.out.versions.first())
```
I  really don't remember why I added it, but it really doesn't make sense to keep it because there is only one emitted output (`CHUNK_FASTA_BUSCO.out.versions`), I will remove it.

alxndrdiaz replied:
Done, removed first from the output versions channel.

alxndrdiaz commented on subworkflows/local/diamond_blastx.nf:
```diff
@@ -0,0 +1,55 @@
+    // Chunk input fasta file
+    //
+    CHUNK_FASTA_BUSCO (
+    fasta.map { fa -> [ [id: fa.baseName ], fa ] },
```
Done.

muffato commented:
@alxndrdiaz : now that #37 is merged, this sub-workflow is next :) Since the branch goes quite far back, maybe make some fresh commits from `dev` ?

alxndrdiaz commented:
> @alxndrdiaz : now that #37 is merged, this sub-workflow is next :) Since the branch goes quite far back, maybe make some fresh commits from `dev` ?

Agree, it would be better to create a new branch from dev.

priyanka-surana commented:
This is an optional subworkflow. Would be better to focus on the required subworkflows first. We can circle back to this at a later point.

priyanka-surana commented:
I would recommend the [`blobtools`](https://github.com/blobtoolkit/blobtoolkit/blob/main/src/blobtoolkit-pipeline/src/blobtools.smk) subworkflow with the [`run_blobtools_create`](https://github.com/blobtoolkit/blobtoolkit/blob/main/src/blobtoolkit-pipeline/src/rules/run_blobtools_create.smk) and [`add_summary_to_metadata`](https://github.com/blobtoolkit/blobtoolkit/blob/main/src/blobtoolkit-pipeline/src/rules/add_summary_to_metadata.smk) modules. I think it might be possible to start with `add_summary_to_metadata` module, since the `run_blobtools_create` requires `window_stats` output which @zb32 is still working on.
